### PR TITLE
chore(func): Add WalkByQuery to Postgres stores

### PR DESCRIPTION
**Pull request by connorgorman**
## Description

Add WalkByQuery to Postgres stores to be used in later export APIs, but just staging these changes first

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

TODO(replace-me)  
Use this space to explain **how you validated** that **your change functions exactly how you expect it**.
Feel free to attach JSON snippets, curl commands, screenshots, etc. Apply a simple benchmark: would the information you
provided convince any reviewer or any external reader that you did enough to validate your change.

It is acceptable to assume trust and keep this section light, e.g. as a bullet-point list.

It is acceptable to skip testing in cases when CI is sufficient, or it's a markdown or code comment change only.  
It is also acceptable to skip testing for changes that are too taxing to test before merging. In such case you are
responsible for the change after it gets merged which includes reverting, fixing, etc. Make sure you validate the change
ASAP after it gets merged or explain in PR when the validation will be performed.  
Explain here why you skipped testing in case you did so.

Have you created automated tests for your change? Explain here which validation activities you did manually and why so.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: pkg/search/postgres/store.go [hunk 1]
```diff
@@ -178,6 +178,32 @@ func (s *GenericStore[T, PT]) Walk(ctx context.Context, fn func(obj PT) error) e
 	return nil
 }
 
+// WalkByQuery iterates over all the objects scoped by the query and applies the closure.
+func (s *GenericStore[T, PT]) WalkByQuery(ctx context.Context, query *v1.Query, fn func(obj PT) error) error {
```
Comment (janisz) [inline pkg/search/postgres/store.go:182]: Can `Walk` be implemented by this?
```go
func (s *GenericStore[T, PT]) Walk(ctx context.Context, fn func(obj PT) error) error {
	return s.WalkWithQuery*ctx, search.EmptyQuery(), fn) 
}
```

Comment (connorgorman) [reply pkg/search/postgres/store.go]: Done, but added an intermediate func so we can distinguish between Walk and WalkByQuery metrics

**Review by dashrews78 — APPROVED** [review]
LGTM

File: central/alert/datastore/datastore_impl.go [hunk 2]
```diff
@@ -252,6 +252,10 @@ func hasSameScope(o1, o2 sac.NamespaceScopedObject) bool {
 	return o1 != nil && o2 != nil && o1.GetClusterId() == o2.GetClusterId() && o1.GetNamespace() == o2.GetNamespace()
 }
 
+func (ds *datastoreImpl) WalkByQuery(ctx context.Context, q *v1.Query, fn func(alert *storage.Alert) error) error {
```
Comment (janisz) [inline central/alert/datastore/datastore_impl.go:255]: Who is using it?

**Review by janisz — APPROVED** [review]

File: central/alert/datastore/datastore_impl.go (hunk 2 above)
Comment (connorgorman) [reply central/alert/datastore/datastore_impl.go:255]: Follow up PRs

**Pull request merged**